File: Algorithm_350_Intersection of Two Arrays II.cpp
#include <iostream>
#include <vector>
using namespace std;
/*vector<int> intersect(vector<int>& nums1, vector<int>& nums2)
{
	int len1=nums1.size(),len2=nums2.size();
	if (len1<len2) return intersect(nums2,nums1);
	vector<int> res;
	if (len1==0 || len2==0) return res;
	bool flag1[len1],flag2[len2];
	for (int i=0;i<len1;i++) flag1[i]=false;
	for (int i=0;i<len2;i++) flag2[i]=false;
	for (int i=0;i<len2;i++)
	{
		for (int j=0;j<len1;j++)
		{
			if (nums2[i]==nums1[j] && !flag1[j] && !flag2[i])
			{
				res.push_back(nums2[i]);
				flag1[j]=true;
				flag2[i]=true;
			}
		}
	}
	return res;
}*/

class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int, int> dict;
        vector<int> res;
        for(int i = 0; i < (int)nums1.size(); i++) dict[nums1[i]]++;
        for(int i = 0; i < (int)nums2.size(); i++)
            if(--dict[nums2[i]] >= 0) res.push_back(nums2[i]);
        return res;
    }
};
int main()
{
	vector<int> n1, n2;
	n1.push_back(1);
	n2.push_back(1);
	n2.push_back(1);
	vector<int> res=intersect(n1,n2);
	for (int i=0;i<res.size();i++) cout<<res[i]<<' ';
	return 0;
}

File: Algorithm_350_Intersection of Two Arrays II.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

// HashMap Solution

class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {
        Map<Integer, Integer> map = new HashMap<>();
        List<Integer> resultList = new ArrayList<>();
        for (int num : nums1) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }
        for (int num : nums2) {
            int count = map.getOrDefault(num, 0);
            if (count > 0) {
                resultList.add(num);
                if (count > 1) {
                    map.put(num, count - 1);
                } else {
                    map.remove(num);
                }
            }
        }
        int[] result = new int[resultList.size()];
        for (int i = 0; i < result.length; i++) {
            result[i] = resultList.get(i);
        }
        return result;
    }
}

// two-pointer solution
class TwoPointerSolution {
    public int[] intersect(int[] nums1, int[] nums2) {
        Arrays.sort(nums1);
        Arrays.sort(nums2);
        int pt1 = 0, pt2 = 0;
        List<Integer> resultList = new ArrayList<>();
        while (pt1 < nums1.length && pt2 < nums2.length) {
            if (nums1[pt1] < nums2[pt2]) {
                pt1++;
            } else if (nums1[pt1] > nums2[pt2]) {
                pt2++;
            } else {
                resultList.add(nums1[pt1]);
                pt1++;
                pt2++;
            }
        }
        int[] result = new int[resultList.size()];
        for (int i = 0; i < result.length; i++) {
            result[i] = resultList.get(i);
        }
        return result;
    }
}

// Follow-up: What if elements of nums2 are stored on disk, and the memory is limited such that you cannot load all elements into the memory at once?
// If only nums2 cannot fit in memory, put all elements of nums1 into a HashMap, read chunks of array that fit into the memory, and record the intersections.
// If both nums1 and nums2 are so huge that neither fit into the memory, sort them individually (external sort), then read 2 elements from each array at a time in memory, 
// record intersections.